gusto/time_discretisation.py

```diff
@@ -1084,3 +1090,495 @@ def apply(self, x_out, *x_in):
+                         options=options)
+
+        self.order = order
+
+    def setup(self, equation, uadv=None, apply_bcs=True, *active_labels):
+        super().setup(equation=equation, uadv=uadv, apply_bcs=apply_bcs,
+                      *active_labels)
+
+        self.x = [Function(self.fs) for i in range(self.nlevels)]
+
+        if (self.order == 1):
+            self.bl = (1.0/2.0)
+            self.br = [(1.0/2.0)]
+        elif (self.order == 2):
+            self.bl = (5.0/12.0)
+            self.br = [-(1.0/12.0), (8.0/12.0)]
+        elif (self.order == 3):
+            self.bl = (9.0/24.0)
+            self.br = [(1.0/24.0), -(5.0/24.0), (19.0/24.0)]
+        elif (self.order == 4):
+            self.bl = (251.0/720.0)
+            self.br = [-(19.0/720.0), (106.0/720.0), -(254.0/720.0), (646.0/720.0)]
+
+    @property
+    def nlevels(self):
+        return self.order
+
+    @property
+    def rhs0(self):
+        """Set up the discretisation's right hand side."""
+        r = self.residual.label_map(
+            all_terms,
+            map_if_true=replace_subject(self.x[-1], self.idx))
+        r = r.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: -0.5*self.dt*t)
+
+        return r.form
+
+    @property
+    def lhs0(self):
+        """Set up the time discretisation's right hand side."""
+        l = self.residual.label_map(
+            all_terms,
+            map_if_true=replace_subject(self.x_out, self.idx))
+        l = l.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: 0.5*self.dt*t)
+        return l.form
+
+    @property
+    def lhs(self):
+        """Set up the time discretisation's right hand side."""
+        l = self.residual.label_map(
+            all_terms,
+            map_if_true=replace_subject(self.x_out, self.idx))
+        l = l.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: self.bl*self.dt*t)
+        return l.form
+
+    @property
+    def rhs(self):
+        """Set up the discretisation's right hand side."""
+        r = self.residual.label_map(all_terms,
+                                    map_if_true=replace_subject(self.x[-1], self.idx))
+        r = r.label_map(lambda t: t.has_label(time_derivative),
+                        map_if_false=lambda t: -self.br[-1]*self.dt*t)
+        for n in range(self.nlevels-1):
+            rtemp = self.residual.label_map(lambda t: t.has_label(time_derivative),
+                                            map_if_true=drop,
+                                            map_if_false=replace_subject(self.x[n], self.idx))
+            rtemp = rtemp.label_map(lambda t: t.has_label(time_derivative),
+                                    map_if_false=lambda t: -self.dt*self.br[n]*t)
+            r += rtemp
+        return r.form
+
+    @property
+    def solver0(self):
+        """Set up the problem and the solver."""
```

Comment:
Could you add a comment to clarify that this is for the first steps? Ideally we should also do this for all the other lhs0/rhs0/solver0 routines

Reply:
I have done this for all schemes where I have added this initial solver and rhs/lhs